pkg/applications/helmclient/client_test.go

```diff
@@ -0,0 +1,411 @@
+	"helm.sh/helm/v3/pkg/uploader"
+
+	kubermaticlog "k8c.io/kubermatic/v2/pkg/log"
+
+	cmdtesting "k8s.io/kubectl/pkg/cmd/testing"
+	"sigs.k8s.io/yaml"
+)
+
+const defaultNs = "default"
+
+func TestHelmClient_NewShouldFailWhenRESTClientGetterNamespaceIsDifferentThanTargetNamespace(t *testing.T) {
+	log := kubermaticlog.New(true, kubermaticlog.FormatJSON).Sugar()
+	downloadDest, settings := createHelmConfigutation(t)
+	defer os.RemoveAll(downloadDest)
+
+	tf := cmdtesting.NewTestFactory().WithNamespace(defaultNs)
+	defer tf.Cleanup()
+
+	_, err := NewClient(tf, settings, "another-ns", log)
+	if err == nil {
+		t.Fatalf("helmclient.NewClient() should fail when RESTClientGetter namespace is different than targetNamespace : %s", err)
+	}
+	expectedErrMsg := "namespace set in RESTClientGetter should be the same as targetNamespace"
+	if !strings.Contains(err.Error(), expectedErrMsg) {
+		t.Fatalf("helmclient.NewClient() fails for the wrong reason. expected error message: '%s' to contains: '%s'", err, expectedErrMsg)
+	}
+}
+
+func TestHelmClient_DownloadChart(t *testing.T) {
+	log := kubermaticlog.New(true, kubermaticlog.FormatJSON).Sugar()
+	srv, err := repotest.NewTempServerWithCleanup(t, "testdata/*.tgz")
+	defer srv.Stop()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ociRegistryUrl := startOciRegistry(t, "testdata/examplechart-0.1.0.tgz")
+
+	testCases := []struct {
+		name         string
+		repoUrl      string
+		chartName    string
+		chartVersion string
+		wantErr      bool
+	}{
+		{
+			name:         "Download from HTTPshould be successful",
+			repoUrl:      srv.URL(),
+			chartName:    "examplechart",
+			chartVersion: "0.1.0",
+			wantErr:      false,
+		},
+		{
+			name:         "Download from HTTP repository should fail when chart does not exist",
+			repoUrl:      srv.URL(),
+			chartName:    "chartthatdoesnotexist",
+			chartVersion: "0.1.0",
+			wantErr:      true,
+		},
+
+		{
+			name:         "Download from OCI repository should be successful",
+			repoUrl:      ociRegistryUrl,
+			chartName:    "examplechart",
+			chartVersion: "0.1.0",
+			wantErr:      false,
+		},
+		{
+			name:         "Download from oci repository should fail when chart does not exist",
+			repoUrl:      ociRegistryUrl,
+			chartName:    "chartthatdoesnotexist",
+			chartVersion: "0.1.0",
+			wantErr:      true,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			func() {
+				downloadDest, settings := createHelmConfigutation(t)
+				defer os.RemoveAll(downloadDest)
+
+				tf := cmdtesting.NewTestFactory().WithNamespace(defaultNs)
+				defer tf.Cleanup()
+
+				helmClient, err := NewClient(tf, settings, defaultNs, log)
+				if err != nil {
+					t.Fatalf("can not init helm Client: %s", err)
+				}
+
+				chartLoc, err := helmClient.DownloadChart(tc.repoUrl, tc.chartName, tc.chartVersion, downloadDest)
+
+				if (err != nil) != tc.wantErr {
+					t.Fatalf("DownloadChart() error = %v, wantErr %v", err, tc.wantErr)
```

Comment:
I am not entirely sure, but why are fataling here? Wouldn't this just be a normal Error? I mean yes we would finish this test case, but we would continue with the others, right?

Reply:
if the download failed, the next assertions will fail too. To avoid confusion message we stop this test case. Other test cases will still be launched even if one fails with fatal.

Reply:
understood